nuxeo-mongodb-directory/src/main/java/org/nuxeo/directory/mongodb/MongoDBReference.java

```diff
@@ -236,35 +238,38 @@ private void setIdsFor(String field, String value, String fieldToUpdate, List<St
 
         List<String> existingIds = getIdsFor(field, value, fieldToUpdate, session);
         for (String id : existingIds) {
-            if (idsToAdd.contains(id)) {
-                idsToAdd.remove(id);
-            } else {
+            if (!idsToAdd.remove(id)) {
                 idsToDelete.add(id);
             }
         }
 
         if (!idsToDelete.isEmpty()) {
+            BasicDBList list = new BasicDBList();
             if (sourceField.equals(field)) {
-                idsToDelete.forEach(id -> removeLink(value, id, session));
+                list.addAll(idsToDelete.stream().map(id -> buildDoc(value, id)).collect(Collectors.toList()));
             } else {
-                idsToDelete.forEach(id -> removeLink(id, value, session));
+                list.addAll(idsToDelete.stream().map(id -> buildDoc(id, value)).collect(Collectors.toList()));
             }
+            Bson deleteDoc = new BasicDBObject("$or", list);
```

Comment:
Maybe you could use `Filters.or(...)`from MongoDB API.
You could also declare `list` as `List<Document>` like it's done below.